### PR TITLE
Add parent pid/comm

### PR DESCRIPTION
**Pull request by TrappM**
We would like to have the parent info in the logs.

sysctl\[3319](sapconf\[3192]) changed kernel/shmmni from 4096 to 32768

In situations when sysctl is used to change a parameter, the caller info might be more interesting than the callee info of sysctl.

### CODE REVIEW

**Comment by Werkov** [comment]
`task->real_parent` is an RCU pointer whose naked dereference doesn't seem safe (due to reparenting).
This would need [bpf_rcu_read_lock](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=9bb00b2895cbfe0ad410457b605d0a72524168c1) enclosure.

**Comment by shunghsiyu** [comment]
Very much agree that displaying parent pid/comm is a needed feature.

@TrappM could you see if it is possible to switch the implementation to use [bpf_rcu_read_lock](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=9bb00b2895cbfe0ad410457b605d0a72524168c1) for accessing `real_parent` pointer as suggested?

**Comment by ghost** [comment]
Makes sense, I will add bpf_rcu_read_lock/unlock but this will need an _#ifdef HAVE_BPF_RCU_READ_LOCK_ as the functions depend on [kernel 6.2]( https://elixir.bootlin.com/linux/v6.2/A/ident/bpf_rcu_read_lock).
Like with the HAVE_CGROUP_CURRENT_FUNC_PROTO, this can be part of the rpm Makefile, right?

**Comment by shunghsiyu** [comment]
> Like with the HAVE_CGROUP_CURRENT_FUNC_PROTO, this can be part of the rpm Makefile, right?

Yes. To be honest I'm not that skilled with Makefile and build system in general, so I'm also open to other approach of managing these feature flags.

Note: now you mentioned this, I realize I forgot to update the main branch to v0.4, and just updated it now, hence the conflict. Sorry about that.

**Comment by shunghsiyu** [comment]
[bpf_rcu_read_lock](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=9bb00b2895cbfe0ad410457b605d0a72524168c1) is only available since v6.2 mainline. So it will only be available on (the SLES equivalent of) Leap 15.6, which is still undergoing development.

@Werkov For earlier kernel could we just accept that there's some chance of reading a stale `real_parent` pointer value and get unless PID/comm logged into the journal? Or it is better to simply leave this feature out on earlier kernel?
(RCU is pure magic to me)

**Comment by ghost** [comment]
switched to KERNEL_VERSION macro for kernel version check and rebased PR to 6f196c4.
just let me know in case you prefer a squashed commit.

**Comment by shunghsiyu** [comment]
TLDR: turns out we don't need [bpf_rcu_read_lock](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=9bb00b2895cbfe0ad410457b605d0a72524168c1) after all. So I've went ahead and merged https://github.com/TrappM/sysctl-logger/commit/a0ee0a4af88f8250ea1f7c607d064f5a3cb99a57

I had previously mislead @Werkov about whether or not BPF program are run with RCU lock. Here's the correct version:
- BPF programs are run under the `rcu_read_lock()`, with the notable exceptions being [sleepable BPF programs](https://github.com/torvalds/linux/commit/1e6c62a88215). We're not setting the `BPF_F_SLEEPABLE` here, for one, but more importantly `BPF_PROG_TYPE_CGROUP_SYSCTL` can't be sleepable, and [it is always run with `rcu_read_lock()` taken](https://github.com/torvalds/linux/blob/v6.6/kernel/bpf/cgroup.c#L1723-L1727).

Apologies to @TrappM for having you go full circle just to come back to the starting point.

**Comment by shunghsiyu** [comment]
I suppose GitHub did not mark this as merged because I had not taken all the commits in the `parent-info`, hmm

**Comment by ghost** [comment]
NP, thanks for merging the relevant changes!
